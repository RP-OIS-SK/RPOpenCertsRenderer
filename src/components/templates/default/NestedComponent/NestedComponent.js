import React, { useState } from "react";
import PropTypes from "prop-types";
import { borders, background } from "./NestedComponent.module.scss";

export const KeyValueComponent = ({ label, value }) => {
  const [show, setShow] = useState(true);
  const toggleShow = () => setShow(!show);
  return (
    <div key={label} className={borders}>
      <div
        className={`p-4 ${background}`}
        onClick={toggleShow}
        data-component-id="toggle-display"
      >
        {label}
      </div>
      <div
        className={`p-4 text-truncate ${show ? "" : "d-none"}`}
        data-component-id="displayed-value"
      >
        {NestedComponent({ children: value })}
      </div>
    </div>
  );
};

const NestedComponent = ({ children }) => {
  if (typeof children !== "object") {
    return children.toString();
  } else {
    const grandchildren = Object.keys(children).map(key =>
      KeyValueComponent({ label: key, value: children[key] })
    );
    return <div className={borders}>{grandchildren}</div>;
  }
};

export default NestedComponent;

NestedComponent.propTypes = {
  children: PropTypes.any
};

KeyValueComponent.propTypes = {
  label: PropTypes.string,
  value: PropTypes.any
};
