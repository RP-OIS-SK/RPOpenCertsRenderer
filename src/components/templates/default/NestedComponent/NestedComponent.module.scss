.borders {
  border-width: 1px;
  border-style: solid;
  border-color: #d5effc;
}

.background {
  background-color: #d5effc;
}
