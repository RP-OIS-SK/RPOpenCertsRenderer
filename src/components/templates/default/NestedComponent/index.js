import NestedComponent from "./NestedComponent";

export default NestedComponent;
