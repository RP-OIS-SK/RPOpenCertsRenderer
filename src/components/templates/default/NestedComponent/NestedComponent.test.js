import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import NestedComponent, { KeyValueComponent } from "./NestedComponent";

describe("KeyValueComponent", () => {
  it("toggles between show and hide value when clicked", () => {
    const component = shallow(<KeyValueComponent value="SOME_TEXT" />);
    const clickableSelector = "[data-component-id='toggle-display']";
    const displayedValueSelector = "[data-component-id='displayed-value']";

    // Default behaviour to display value
    const initialClasses = component.find(displayedValueSelector).props()
      .className;
    expect(initialClasses).not.toContain("d-none");

    // Hide value when label is clicked on
    component.find(clickableSelector).simulate("click");
    const intermediateClasses = component.find(displayedValueSelector).props()
      .className;
    expect(intermediateClasses).toContain("d-none");

    // Show value again when label is clicked
    component.find(clickableSelector).simulate("click");
    const finalClasses = component.find(displayedValueSelector).props()
      .className;
    expect(finalClasses).not.toContain("d-none");
  });
});

describe("NestedComponent", () => {
  it("renders primitives", () => {
    const stringPrimitive = renderer.create(
      <NestedComponent>{"Hello"}</NestedComponent>
    );
    expect(stringPrimitive.toJSON()).toMatchSnapshot();

    const numberPrimitive = renderer.create(
      <NestedComponent>{1337}</NestedComponent>
    );
    expect(numberPrimitive.toJSON()).toMatchSnapshot();

    const booleanPrimitive = renderer.create(
      <NestedComponent>{false}</NestedComponent>
    );
    expect(booleanPrimitive.toJSON()).toMatchSnapshot();
  });

  it("renders an array with KeyValueComponent", () => {
    const sampleArray = ["ITEM_1", "ITEM_2", "ITEM_3"];
    const component = renderer.create(
      <NestedComponent>{sampleArray}</NestedComponent>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders an object with KeyValueComponent", () => {
    const sampleObject = { foo: "bar", cow: "moo" };
    const component = renderer.create(
      <NestedComponent>{sampleObject}</NestedComponent>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders a nested object with KeyValueComponent", () => {
    const sampleObject = {
      foo: "bar",
      cow: { nested: "moo" },
      array: ["TEXT_1", 2, false]
    };
    const component = renderer.create(
      <NestedComponent>{sampleObject}</NestedComponent>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
