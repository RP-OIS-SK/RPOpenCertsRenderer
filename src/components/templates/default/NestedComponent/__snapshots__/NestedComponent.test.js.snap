// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedComponent renders a nested object with KeyValueComponent 1`] = `
<div
  className="borders"
>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      foo
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      bar
    </div>
  </div>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      cow
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      <div
        className="borders"
      >
        <div
          className="borders"
        >
          <div
            className="p-4 background"
            data-component-id="toggle-display"
            onClick={[Function]}
          >
            nested
          </div>
          <div
            className="p-4 text-truncate "
            data-component-id="displayed-value"
          >
            moo
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      array
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      <div
        className="borders"
      >
        <div
          className="borders"
        >
          <div
            className="p-4 background"
            data-component-id="toggle-display"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="p-4 text-truncate "
            data-component-id="displayed-value"
          >
            TEXT_1
          </div>
        </div>
        <div
          className="borders"
        >
          <div
            className="p-4 background"
            data-component-id="toggle-display"
            onClick={[Function]}
          >
            1
          </div>
          <div
            className="p-4 text-truncate "
            data-component-id="displayed-value"
          >
            2
          </div>
        </div>
        <div
          className="borders"
        >
          <div
            className="p-4 background"
            data-component-id="toggle-display"
            onClick={[Function]}
          >
            2
          </div>
          <div
            className="p-4 text-truncate "
            data-component-id="displayed-value"
          >
            false
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NestedComponent renders an array with KeyValueComponent 1`] = `
<div
  className="borders"
>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      0
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      ITEM_1
    </div>
  </div>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      ITEM_2
    </div>
  </div>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      2
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      ITEM_3
    </div>
  </div>
</div>
`;

exports[`NestedComponent renders an object with KeyValueComponent 1`] = `
<div
  className="borders"
>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      foo
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      bar
    </div>
  </div>
  <div
    className="borders"
  >
    <div
      className="p-4 background"
      data-component-id="toggle-display"
      onClick={[Function]}
    >
      cow
    </div>
    <div
      className="p-4 text-truncate "
      data-component-id="displayed-value"
    >
      moo
    </div>
  </div>
</div>
`;

exports[`NestedComponent renders primitives 1`] = `"Hello"`;

exports[`NestedComponent renders primitives 2`] = `"1337"`;

exports[`NestedComponent renders primitives 3`] = `"false"`;
